[SPMD] Add the default graph module transformation that is applied after tracing and expansion

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #98580
* #98579
* #98578
* #98285
* #98284
* __->__ #98182

This PR adds the GraphModuleTransformation class that can be used as the
default transformation after the `train_step()` is traced and expand. The
current implementation includes:
1. Wrap the input graph module with IterGraphModule. This will enable the futher graph optimizations which are all implemented based on IterGraphModule.
2. Ability to lower the graph module to the Inductor. To achieve this goal, `lower_to_inductor()` is implemented.

TODO:
1. The `override` and `gm_transofmation` have overlapping functions -- `override.transform` can be used to achieve the same function as `gm_transformation`. However, the current semantics of `override` is to override and transform partial graphs while `gm_transformation` is to transform the entire expaned GM. The final UX of `compile()` needs some discussion.

2. The current `lower_to_inductor()` assumes that the entire graph can be lowered to Inductor. This assumption is okay for integration of graph optimizations but is too restrictive for many models. We should upstream `partial_lowering()`.

Differential Revision: [D44616783](https://our.internmc.facebook.com/intern/diff/D44616783/)

fegin committed:
[SPMD] Add the default graph module transformation that is applied after tracing and expansion

This PR adds the GraphModuleTransformation class that can be used as the
default transformation after the `train_step()` is traced and expand. The
current implementation includes:
1. Wrap the input graph module with IterGraphModule. This will enable the futher graph optimizations which are all implemented based on IterGraphModule.
2. Ability to lower the graph module to the Inductor. To achieve this goal, `lower_to_inductor()` is implemented.

TODO:
1. The `override` and `gm_transofmation` have overlapping functions -- `override.transform` can be used to achieve the same function as `gm_transformation`. However, the current semantics of `override` is to override and transform partial graphs while `gm_transformation` is to transform the entire expaned GM. The final UX of `compile()` needs some discussion.

2. The current `lower_to_inductor()` assumes that the entire graph can be lowered to Inductor. This assumption is okay for integration of graph optimizations but is too restrictive for many models. We should upstream `partial_lowering()`.

Differential Revision: [D44616783](https://our.internmc.facebook.com/intern/diff/D44616783/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Add the default graph module transformation that is applied after tracing and expansion"


This PR adds the GraphModuleTransformation class that can be used as the
default transformation after the `train_step()` is traced and expand. The
current implementation includes:
1. Wrap the input graph module with IterGraphModule. This will enable the futher graph optimizations which are all implemented based on IterGraphModule.
2. Ability to lower the graph module to the Inductor. To achieve this goal, `lower_to_inductor()` is implemented.

TODO:
1. The `override` and `gm_transofmation` have overlapping functions -- `override.transform` can be used to achieve the same function as `gm_transformation`. However, the current semantics of `override` is to override and transform partial graphs while `gm_transformation` is to transform the entire expaned GM. The final UX of `compile()` needs some discussion.

2. The current `lower_to_inductor()` assumes that the entire graph can be lowered to Inductor. This assumption is okay for integration of graph optimizations but is too restrictive for many models. We should upstream `partial_lowering()`.

Differential Revision: [D44616783](https://our.internmc.facebook.com/intern/diff/D44616783/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Add the default graph module transformation that is applied after tracing and expansion"


This PR adds the GraphModuleTransformation class that can be used as the
default transformation after the `train_step()` is traced and expand. The
current implementation includes:
1. Wrap the input graph module with IterGraphModule. This will enable the futher graph optimizations which are all implemented based on IterGraphModule.
2. Ability to lower the graph module to the Inductor. To achieve this goal, `lower_to_inductor()` is implemented.

TODO:
1. The `override` and `gm_transofmation` have overlapping functions -- `override.transform` can be used to achieve the same function as `gm_transformation`. However, the current semantics of `override` is to override and transform partial graphs while `gm_transformation` is to transform the entire expaned GM. The final UX of `compile()` needs some discussion.

2. The current `lower_to_inductor()` assumes that the entire graph can be lowered to Inductor. This assumption is okay for integration of graph optimizations but is too restrictive for many models. We should upstream `partial_lowering()`.

Differential Revision: [D44616783](https://our.internmc.facebook.com/intern/diff/D44616783/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Add the default graph module transformation that is applied after tracing and expansion"


This PR adds the GraphModuleTransformation class that can be used as the
default transformation after the `train_step()` is traced and expand. The
current implementation includes:
1. Wrap the input graph module with IterGraphModule. This will enable the futher graph optimizations which are all implemented based on IterGraphModule.
2. Ability to lower the graph module to the Inductor. To achieve this goal, `lower_to_inductor()` is implemented.

TODO:
1. The `override` and `gm_transofmation` have overlapping functions -- `override.transform` can be used to achieve the same function as `gm_transformation`. However, the current semantics of `override` is to override and transform partial graphs while `gm_transformation` is to transform the entire expaned GM. The final UX of `compile()` needs some discussion.

2. The current `lower_to_inductor()` assumes that the entire graph can be lowered to Inductor. This assumption is okay for integration of graph optimizations but is too restrictive for many models. We should upstream `partial_lowering()`.

Differential Revision: [D44616783](https://our.internmc.facebook.com/intern/diff/D44616783/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Add the default graph module transformation that is applied after tracing and expansion"


This PR adds the GraphModuleTransformation class that can be used as the
default transformation after the `train_step()` is traced and expand. The
current implementation includes:
1. Wrap the input graph module with IterGraphModule. This will enable the futher graph optimizations which are all implemented based on IterGraphModule.
2. Ability to lower the graph module to the Inductor. To achieve this goal, `lower_to_inductor()` is implemented.

TODO:
1. The `override` and `gm_transofmation` have overlapping functions -- `override.transform` can be used to achieve the same function as `gm_transformation`. However, the current semantics of `override` is to override and transform partial graphs while `gm_transformation` is to transform the entire expaned GM. The final UX of `compile()` needs some discussion.

2. The current `lower_to_inductor()` assumes that the entire graph can be lowered to Inductor. This assumption is okay for integration of graph optimizations but is too restrictive for many models. We should upstream `partial_lowering()`.

Differential Revision: [D44616783](https://our.internmc.facebook.com/intern/diff/D44616783/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Add the default graph module transformation that is applied after tracing and expansion"


This PR adds the GraphModuleTransformation class that can be used as the
default transformation after the `train_step()` is traced and expand. The
current implementation includes:
1. Wrap the input graph module with IterGraphModule. This will enable the futher graph optimizations which are all implemented based on IterGraphModule.
2. Ability to lower the graph module to the Inductor. To achieve this goal, `lower_to_inductor()` is implemented.

TODO:
1. The `override` and `gm_transofmation` have overlapping functions -- `override.transform` can be used to achieve the same function as `gm_transformation`. However, the current semantics of `override` is to override and transform partial graphs while `gm_transformation` is to transform the entire expaned GM. The final UX of `compile()` needs some discussion.

2. The current `lower_to_inductor()` assumes that the entire graph can be lowered to Inductor. This assumption is okay for integration of graph optimizations but is too restrictive for many models. We should upstream `partial_lowering()`.

Differential Revision: [D44616783](https://our.internmc.facebook.com/intern/diff/D44616783/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Add the default graph module transformation that is applied after tracing and expansion"


This PR adds the GraphModuleTransformation class that can be used as the
default transformation after the `train_step()` is traced and expand. The
current implementation includes:
1. Wrap the input graph module with IterGraphModule. This will enable the futher graph optimizations which are all implemented based on IterGraphModule.
2. Ability to lower the graph module to the Inductor. To achieve this goal, `lower_to_inductor()` is implemented.

TODO:
1. The `override` and `gm_transofmation` have overlapping functions -- `override.transform` can be used to achieve the same function as `gm_transformation`. However, the current semantics of `override` is to override and transform partial graphs while `gm_transformation` is to transform the entire expaned GM. The final UX of `compile()` needs some discussion.

2. The current `lower_to_inductor()` assumes that the entire graph can be lowered to Inductor. This assumption is okay for integration of graph optimizations but is too restrictive for many models. We should upstream `partial_lowering()`.

Differential Revision: [D44616783](https://our.internmc.facebook.com/intern/diff/D44616783/)

[ghstack-poisoned]

wanchaol commented on torch/distributed/_spmd/gm_transformation.py:
```diff
@@ -0,0 +1,131 @@
+    return gm
+
+
+class GraphModuleTransformation:
```
is it possible to just embed this to the IterGraphModule itself? I think it's better to have less nesting abstractions if not necessary.

fegin replied:
It can. Will do this later and integrate IterGraphModule with Inductor. After that this class will be come a function call.

wanchaol commented on torch/distributed/_spmd/gm_transformation.py:
```diff
@@ -0,0 +1,131 @@
+    def __call__(self, gm: fx.GraphModule) -> Callable:
+        iter_gm = IterGraphModule(gm)
+        iter_gm.freeze_cross_iter_movement()
+        iter_gm.setup(self.num_iters)
```
could these two methods be called automatically in `IterGraphModule` constructor?

fegin replied:
`freeze_cross_iter_movement()` can be called after IterGraphModule is integrated with Inductor.  `setup(num_iters)` is not possible to be removed at this moment.

Some examples that extra function call or extra argument to `forward()` is required other than simply arguments to the constructor:

1. Multiple epochs. For multiple epoch case, we need some signal to indicate that the current training is over and a reset will be needed for the next training loop.
2. Data depleted, similar to example 1. We need a way for users to let IterGraphModule know that training is over.

Will need some discussion about what's the easiest way for users to inform this information. Right now, we have to keep this API.

mrshenli reviewed (commented):
> The override and gm_transofmation have overlapping functions -- override.transform can be used to achieve the same function as gm_transformation. However, the current semantics of override is to override and transform partial graphs while gm_transformation is to transform the entire expaned GM. The final UX of compile() needs some discussion.

We can drop the transform in Override and have replacement and transform as two top level args under compile?

mrshenli commented on torch/distributed/_spmd/gm_transformation.py:
```diff
@@ -0,0 +1,131 @@
+        node_mapping[node] = subgraph.create_node(
+            op=node.op, target=node.target, args=args, kwargs=kwargs
+        )
+        node_mapping[node].meta = node.meta
```
`node_copy` seems handles both arg remap and .meta field

fegin replied:
good point. We should use `node_copy` here.

mrshenli commented on torch/distributed/_spmd/gm_transformation.py:
```diff
@@ -0,0 +1,131 @@
+        new_placeholder = node_mapping[placeholder]
+        if len(new_placeholder.users) == 0:
+            # Remove unused placeholders from the subgraph. This is required as
+            # the `train_step()` has module and optimizer as the inputs which
```
(no need to change this PR) in this case, should we do this right after getting the local train_step graph instead of dtensor expansion and other graph transformations?

fegin replied:
Users will still pass model and optimizer, right? So do you mean that we directly wrap the FX graph right after expansion?

mrshenli replied:
> No, I was referring to before expansion. 

Basically, trace local graph -> remove unused placeholders -> DTensor expansion

Because expansion is graph-traversal based solution for now, and does not need to pass in models and optimziers. 

But @wanchaol's new SPMD expansion might potentially change that, suppose we do it through our dedicated tracer.

fegin replied:
Still a little confused about this. A user has to call `train_step(model, optim, batch)` to trace the single step and expand the graph. The user will expect that the subsequent calls are still `train_step(model, optim, batch)` which means `model` and `optim` are the placeholders for the `fx.Graph`. We have to create a subgraph in order to drop `model` and `optim`. That subgraph can then be lowered to the Inductor. We can do this wrapping during or right after expansion which makes sense to me. But I want to confirm my understanding that the wrapping, creating a subgraph, is still needed?

mrshenli commented on torch/distributed/_spmd/gm_transformation.py:
```diff
@@ -0,0 +1,131 @@
+        else:
+            placeholders.append(placeholder)
+    output_args = tuple(
+        node_mapping[n] for n in orig_output_args if n not in orig_placeholders_set
```
curious, do we see cases where output args contains original inputs?

fegin replied:
Yes, the tracing make parameters and opt_state as input and also the updated parameters and opt_state as output. However, there can be cases where those inputs are not updated. For example, the step is not called.

fegin replied:
Yes, it is possible. But that is handled below instead of passing the placeholder into the subgraph and redirect it to subgraph output again.

fegin replied:
Updated on this. I decided to remove the placeholder check and redirect all the output even if it is directly from placeholder. The logic is cleaner.

mrshenli commented on torch/distributed/_spmd/gm_transformation.py:
```diff
@@ -0,0 +1,131 @@
+    )
+    subgraph.output(result=output_args)
+    sub_gm = torch.fx.GraphModule(root=attrs, graph=subgraph)
+    gm.subgraph = InductorWrapper(sub_gm, enable_cudagraphs)
```
curious, what happens to our comm nodes? Will inductor's horizontal fusion respect comm nodes as boundaries?

fegin replied:
No, it is possible that Inductor's horizontal fusion may break the boundaries and delay the comm nodes. @lessw2020 has done experiments show that the current fusion algorithm won't cause this issue. But there is no guarantee this issue won't happen. We have to add hints for Inductor to not fusion across the comm_ops.

lessw2020 replied:
We still have the code I wrote earlier to implement 'comm pools' which forces Inductor to respect our comm boundaries (i.e. it's fusions will only take place within the boundaries of comm calls). 
In general, the backward fusions from Inductor span < 8 or so nodes, and so are unlikely to cause issues, but we can setup with this and compare perf and/or it also generates stats on whether Inductor fusion crossed our comm boundaries. (it was not checked in, but lives here: https://github.com/lessw2020/tau_graph/commits/comm_scheduler/spmd/comm_scheduler.py)

fegin replied:
cc., @Chillee and @sanketpurandare

mrshenli commented on torch/distributed/_spmd/gm_transformation.py:
```diff
@@ -0,0 +1,131 @@
+    gm.subgraph = InductorWrapper(sub_gm, enable_cudagraphs)
+    with gm.graph.inserting_after(move_nodes[-1]):
+        subgraph_call = gm.graph.create_node(
+            op=OP.CALL_MODULE, target="subgraph", args=tuple(placeholders)
```
I assume this means except inputs and outpus nodes, everything else will be in this big codegen-ed node?

fegin replied:
yes, this API is meant to be used for full lowering. @yifuwang will upstream partial lowering.

mrshenli commented on torch/distributed/_spmd/gm_transformation.py:
```diff
@@ -0,0 +1,131 @@
+                operator.getitem, (subgraph_call, output_idx)
+            )
+            output_idx += 1
+        orig_output_args[i] = new_node
```
is the order guaranteed to be consistent or is this an internal behavior that we rely on?

fegin replied:
We create the subgraph output_args based on the order of `orig_output_args` in line 80. So the order should be consistent?

mrshenli commented on torch/distributed/_spmd/gm_transformation.py:
```diff
@@ -0,0 +1,131 @@
+def lower_to_inductor(
+    gm: torch.fx.GraphModule, enable_cudagraphs: bool
+) -> torch.fx.GraphModule:
+    orig_placeholders = []
+    orig_output_args: List[Any] = []
```
nit: let's be consistent on typing hint availability

mrshenli commented on test/distributed/_spmd/test_transformation.py:
```diff
@@ -0,0 +1,135 @@
+    def _init(self, batch_size, layers, dim):
+        torch.manual_seed(0)
+        model = DummyModel(layers, dim).cuda()
+        ddp_model = DDP(deepcopy(model), device_ids=[dist.get_rank()])
```
nit: `self.rank` should work

mrshenli commented on test/distributed/_spmd/test_transformation.py:
```diff
@@ -0,0 +1,135 @@
+class DummyModel(nn.Module):
+    def __init__(self, layers: int, dim: int):
+        super().__init__()
+        self.mod = nn.Sequential(*((nn.Linear(dim, dim), nn.ReLU()) * layers))
```
I see, this is where shared params are coming from :)

fegin replied:
Yup, :( I changed it locally, have not updated the PR yet. :

mrshenli commented on test/distributed/_spmd/test_transformation.py:
```diff
@@ -0,0 +1,135 @@
+        ddp_optim.step()
+        ddp_optim.zero_grad()
+
+        self.assertEqual(ddp_out, out)
```
nit: assert on params as well?

fegin commented on torch/distributed/_spmd/gm_transformation.py:
```diff
@@ -0,0 +1,131 @@
+    node_mapping: Dict[torch.fx.Node, torch.fx.Node] = {}
+    attrs = {}
+    for i, p in enumerate(orig_placeholders):
+        node_mapping[p] = subgraph.placeholder(name=f"placeholder_{i}")
```
No. We can use `node_copy`.

fegin committed:
Update on "[SPMD] Add the default graph module transformation that is applied after tracing and expansion"


This PR adds the GraphModuleTransformation class that can be used as the
default transformation after the `train_step()` is traced and expand. The
current implementation includes:
1. Wrap the input graph module with IterGraphModule. This will enable the futher graph optimizations which are all implemented based on IterGraphModule.
2. Ability to lower the graph module to the Inductor. To achieve this goal, `lower_to_inductor()` is implemented.

TODO:
1. The `override` and `gm_transofmation` have overlapping functions -- `override.transform` can be used to achieve the same function as `gm_transformation`. However, the current semantics of `override` is to override and transform partial graphs while `gm_transformation` is to transform the entire expaned GM. The final UX of `compile()` needs some discussion.

2. The current `lower_to_inductor()` assumes that the entire graph can be lowered to Inductor. This assumption is okay for integration of graph optimizations but is too restrictive for many models. We should upstream `partial_lowering()`.

Differential Revision: [D44616783](https://our.internmc.facebook.com/intern/diff/D44616783/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Add the default graph module transformation that is applied after tracing and expansion"


This PR adds the GraphModuleTransformation class that can be used as the
default transformation after the `train_step()` is traced and expand. The
current implementation includes:
1. Wrap the input graph module with IterGraphModule. This will enable the futher graph optimizations which are all implemented based on IterGraphModule.
2. Ability to lower the graph module to the Inductor. To achieve this goal, `lower_to_inductor()` is implemented.

TODO:
1. The `override` and `gm_transofmation` have overlapping functions -- `override.transform` can be used to achieve the same function as `gm_transformation`. However, the current semantics of `override` is to override and transform partial graphs while `gm_transformation` is to transform the entire expaned GM. The final UX of `compile()` needs some discussion.

2. The current `lower_to_inductor()` assumes that the entire graph can be lowered to Inductor. This assumption is okay for integration of graph optimizations but is too restrictive for many models. We should upstream `partial_lowering()`.

Differential Revision: [D44616783](https://our.internmc.facebook.com/intern/diff/D44616783/)

[ghstack-poisoned]

mrshenli commented on test/distributed/_spmd/test_transformation.py:
```diff
@@ -0,0 +1,142 @@
+            batch = torch.randn(batch_size, dim).cuda()
+            out = train_step(model, optim, batch)
+
+            # TODO: using _test_tran_step_with_ddp_without_optim_step()
```
Hey @fegin, do you plan to complete this test before landing, or is this for followup PRs?

fegin replied:
Done in the latest version.

fegin committed:
Update on "[SPMD] Add the default graph module transformation that is applied after tracing and expansion"


This PR adds the GraphModuleTransformation class that can be used as the
default transformation after the `train_step()` is traced and expand. The
current implementation includes:
1. Wrap the input graph module with IterGraphModule. This will enable the futher graph optimizations which are all implemented based on IterGraphModule.
2. Ability to lower the graph module to the Inductor. To achieve this goal, `lower_to_inductor()` is implemented.

TODO:
1. The `override` and `gm_transofmation` have overlapping functions -- `override.transform` can be used to achieve the same function as `gm_transformation`. However, the current semantics of `override` is to override and transform partial graphs while `gm_transformation` is to transform the entire expaned GM. The final UX of `compile()` needs some discussion.

2. The current `lower_to_inductor()` assumes that the entire graph can be lowered to Inductor. This assumption is okay for integration of graph optimizations but is too restrictive for many models. We should upstream `partial_lowering()`.

Differential Revision: [D44616783](https://our.internmc.facebook.com/intern/diff/D44616783/)

[ghstack-poisoned]

mrshenli commented on test/distributed/_spmd/test_transformation.py:
```diff
@@ -0,0 +1,137 @@
+        # make sure we set different random seeds for each rank
+        # otherwise we dont need DDP / SPMD
+        # (we would have the same parameters and inputs everywhere)
+        torch.manual_seed(torch.distributed.get_rank())
```
nit: (can be done in followup PR), it seems `with_comms` always wraps test cases, where `self.rank` should exist.

fegin replied:
Will fix this in a separate PR.

mrshenli commented on torch/distributed/_spmd/gm_transformation.py:
```diff
@@ -0,0 +1,130 @@
+            )
+            output_idx += 1
+        orig_output_args[i] = new_node
+    assert output_idx == len(output_args)
```
nit: curious, why do we need `output_idx` instead of just `len(orig_output_args) == len(output_args)`? It seems `output_idx` is always the same as  `len(orig_output_args)`?

fegin replied:
You are right, this is a legacy from the previous version. Will fix it in a separate PR.

mrshenli reviewed (approved):
LGTM

fegin committed:
Update on "[SPMD] Add the default graph module transformation that is applied after tracing and expansion"


This PR adds the GraphModuleTransformation class that can be used as the
default transformation after the `train_step()` is traced and expand. The
current implementation includes:
1. Wrap the input graph module with IterGraphModule. This will enable the futher graph optimizations which are all implemented based on IterGraphModule.
2. Ability to lower the graph module to the Inductor. To achieve this goal, `lower_to_inductor()` is implemented.

TODO:
1. The `override` and `gm_transofmation` have overlapping functions -- `override.transform` can be used to achieve the same function as `gm_transformation`. However, the current semantics of `override` is to override and transform partial graphs while `gm_transformation` is to transform the entire expaned GM. The final UX of `compile()` needs some discussion.

2. The current `lower_to_inductor()` assumes that the entire graph can be lowered to Inductor. This assumption is okay for integration of graph optimizations but is too restrictive for many models. We should upstream `partial_lowering()`.

Differential Revision: [D44616783](https://our.internmc.facebook.com/intern/diff/D44616783/)

[ghstack-poisoned]

fegin commented:
@pytorchbot merge
